core_impl/src/main/java/io/opencensus/implcore/stats/MutableViewData.java

```diff
@@ -283,6 +289,167 @@ ViewData toViewData(Clock clock) {
+    private final Duration totalDuration;  // Duration of the Interval window
+    private final Duration bucketDuration; // Duration of a single bucket (totalDuration / N)
+
+    private Timestamp startOfBuckets; // Start timestamp of the eldest bucket of all current buckets
+
+    private IntervalMutableViewData(View view, Timestamp start) {
+      super(view);
+      totalDuration = ((Interval) view.getWindow()).getDuration();
+      bucketDuration = Duration.fromMillis(toMillis(totalDuration) / (MAX_BUCKET_SIZE - 1));
+
+      // When initializing. add N empty buckets prior to the start timestamp of this
+      // IntervalMutableViewData, so that the last bucket will be the current one in effect.
+      startOfBuckets = start.addDuration(
+          Duration.create(-totalDuration.getSeconds(), -totalDuration.getNanos()));
+      refreshBucketList(MAX_BUCKET_SIZE);
+    }
+
+    @Override
+    void record(TagContext context, double value, Timestamp timestamp) {
+      List<TagValue> tagValues = getTagValues(getTagMap(context), super.view.getColumns());
+      padBuckets(timestamp);
+      // It is always the last bucket that does the recording.
+      buckets.peekLast().record(tagValues, value);
+    }
+
+    @Override
+    void record(TagContext tags, long value, Timestamp timestamp) {
+      // TODO(songya): implement this.
+      throw new UnsupportedOperationException("Not implemented.");
+    }
+
+    @Override
+    ViewData toViewData(Clock clock) {
+      Timestamp now = clock.now();
+      padBuckets(now);
+      return ViewData.create(
+          super.view, combineBucketsAndGetAggregationMap(now), IntervalData.create(now));
+    }
+
+    // Add new buckets and remove expired buckets by comparing the current timestamp with
+    // timestamp of the last bucket.
+    private void padBuckets(Timestamp now) {
+      checkArgument(buckets.size() == MAX_BUCKET_SIZE,
+          "Invariant 1: bucket list must have exactly " + MAX_BUCKET_SIZE + " buckets.");
+      Timestamp startOfLastBucket = buckets.peekLast().getStart();
+      checkArgument(now.compareTo(startOfLastBucket) >= 0,
+          "Invariant 2: current time must be within or after the last bucket.");
+      long elapsedTimeMillis = toMillis(now.subtractTimestamp(startOfLastBucket));
+      long numOfPadBuckets = elapsedTimeMillis / toMillis(bucketDuration);
+
+      refreshBucketList(numOfPadBuckets);
+    }
+
+    // Add specified number of new buckets, and remove expired buckets
+    private void refreshBucketList(long numOfPadBuckets) {
+      Timestamp startOfNewBucket;
+      if (!buckets.isEmpty()) {
+        startOfNewBucket = buckets.peekLast().getStart().addDuration(bucketDuration);
+      } else {
+        // Initialize bucket list. Should only enter this block once.
+        startOfNewBucket = startOfBuckets;
+      }
+
+      if (numOfPadBuckets > MAX_BUCKET_SIZE) {
+        // All current buckets expired, need to add N + 1 new buckets, and update startOfNewBucket.
```

Comment:
I think there should be a test for this case, i.e., record more stats after all initial stats have expired.